primitives/npos-elections/src/pjr.rs

```diff
@@ -133,18 +133,103 @@ pub fn t_pjr_check<AccountId: IdentifierT>(
 /// The internal implementation of the PJR check after having the data converted.
 ///
 /// [`pjr_check`] or [`t_pjr_check`] are typically easier to work with.
+///
+/// This function returns an `AccountId` in the `Err` case. This is the counter_example: the ID of the
+/// unelected candidate with the highest prescore, such that `pre_score(counter_example) >= t`.
 pub fn pjr_check_core<AccountId: IdentifierT>(
 	candidates: &[CandidatePtr<AccountId>],
 	voters: &[Voter<AccountId>],
 	t: Threshold,
-) -> bool {
+) -> Result<(), AccountId> {
 	let unelected = candidates.iter().filter(|c| !c.borrow().elected);
 	let maybe_max_pre_score = unelected.map(|c| (pre_score(Rc::clone(c), voters, t), c.borrow().who.clone())).max();
 	// if unelected is empty then the solution is indeed PJR.
-	maybe_max_pre_score.map_or(true, |(max_pre_score, _)| max_pre_score < t)
+	match maybe_max_pre_score {
+		None => Ok(()),
+		Some((max_pre_score, counter_example)) => if max_pre_score < t {
+			Ok(())
+		} else {
+			Err(counter_example)
+		}
+	}
 }
 
+/// Validate a challenge to an election result.
+///
+/// A challenge to an election result is valid if there exists some counter_example for which
+/// `pre_score(counter_example) >= threshold`. Validating an existing counter_example is computationally
+/// cheaper than re-running the PJR check.
+///
+/// This function uses the standard threshold.
+///
+/// Returns `true` if the challenge is valid: the proposed solution does not satisfy PJR.
+/// Returns `false` if the challenge is invalid: the proposed solution does in fact satisfy PJR.
+pub fn validate_pjr_challenge<AccountId: IdentifierT>(
+	counter_example: AccountId,
+	supports: &Supports<AccountId>,
+	all_candidates: Vec<AccountId>,
+	all_voters: Vec<(AccountId, VoteWeight, Vec<AccountId>)>,
```

Comment:
One random note unrelated to this PR per se: this pallet had quite a bad rap for not having a well defined behaviour with duplicate votes and candidates and targets etc. Now, finally, everything that goes through `setup_input` (called prior to `seq_phragmen` and `phragmms`) gets sanitized. Would be good to ensure these functions have the same standard.

Reply:
[`prepare_pjr_input`](https://github.com/paritytech/substrate/blob/2d695630e3eabfa9b0b2be8d8e19f16dfe70e323/primitives/npos-elections/src/pjr.rs#L230-L261) mentions `setup_inputs` in its documentation, but doesn't actually delegate. Shall I create an issue to add explicit delegation there to ensure that the behaviors are identical?

Reply:
An issue would be good, bit I am not sure if delegation is the best way, we have to see. We can maybe mix this into #4593